src/search/pdbs/pattern_database.cc

```diff
@@ -1,372 +1,42 @@
-          Note that this kind of plan extraction does not uniformly at random
-          consider all successor of a state but rather uses the arbitrarily
-          chosen generating operator to settle on one successor state, which
-          is biased by the number of operators leading to the same successor
-          from the given state.
-        */
-        State initial_state = task_proxy.get_initial_state();
-        initial_state.unpack();
-        int current_state =
-            hash_index(initial_state.get_unpacked_values());
-        if (distances[current_state] != numeric_limits<int>::max()) {
-            while (!is_goal_state(current_state, abstract_goals, variables)) {
-                int op_id = generating_op_ids[current_state];
-                assert(op_id != -1);
-                const AbstractOperator &op = operators[op_id];
-                int successor_state = current_state - op.get_hash_effect();
-
-                // Compute equivalent ops
-                vector<OperatorID> cheapest_operators;
-                vector<int> applicable_operator_ids;
-                match_tree.get_applicable_operator_ids(successor_state, applicable_operator_ids);
-                for (int applicable_op_id : applicable_operator_ids) {
-                    const AbstractOperator &applicable_op = operators[applicable_op_id];
-                    int predecessor = successor_state + applicable_op.get_hash_effect();
-                    if (predecessor == current_state && op.get_cost() == applicable_op.get_cost()) {
-                        cheapest_operators.emplace_back(applicable_op.get_concrete_op_id());
-                    }
-                }
-                if (compute_wildcard_plan) {
-                    rng->shuffle(cheapest_operators);
-                    wildcard_plan.push_back(move(cheapest_operators));
-                } else {
-                    OperatorID random_op_id = *rng->choose(cheapest_operators);
-                    wildcard_plan.emplace_back();
-                    wildcard_plan.back().push_back(random_op_id);
-                }
-
-                current_state = successor_state;
-            }
-        }
-        utils::release_vector_memory(generating_op_ids);
+        index += hash_multipliers[i] * state[pattern[i]];
     }
+    return index;
 }
 
-bool PatternDatabase::is_goal_state(
-    int state_index,
-    const vector<FactPair> &abstract_goals,
-    const VariablesProxy &variables) const {
-    for (const FactPair &abstract_goal : abstract_goals) {
-        int pattern_var_id = abstract_goal.var;
-        int var_id = pattern[pattern_var_id];
-        VariableProxy var = variables[var_id];
-        int temp = state_index / hash_multipliers[pattern_var_id];
-        int val = temp % var.get_domain_size();
-        if (val != abstract_goal.value) {
-            return false;
-        }
-    }
-    return true;
+int Projection::unrank(int index, int var, int domain_size) const {
```

Comment:
Projection could store the list of domain sizes of the variables in pattern. However, unranking is only required in MatchTree and the distance computation. Match tree nodes store the domain size to know how many successors they have. Since they use arrays, they need that information for their destruction as well. So, storing the domain sizes for the pattern variables would waste some memory once the PDB is computed, and the match tree would need to rely on the fact that the projection has the same domain sizes as those that it uses itself. I'm still not really opposed to the suggestion, as we recently implemented domain abstractions and there we did exactly that: store the domain sizes of all "pattern" variables. Do you think this is a valuable change?